##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# Text classification with TensorFlow Hub: Movie reviews

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/tutorials/keras/text_classification_with_hub"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View on GitHub</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
  <td>
    <a href="https://tfhub.dev/s?module-type=text-embedding"><img src="https://www.tensorflow.org/images/hub_logo_32px.png" />See TF Hub models</a>
  </td>
</table>

This notebook classifies movie reviews as *positive* or *negative* using the text of the review. This is an example of *binary*—or two-class—classification, an important and widely applicable kind of machine learning problem.

The tutorial demonstrates the basic application of transfer learning with [TensorFlow Hub](https://tfhub.dev) and Keras.

We'll use the [IMDB dataset](https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb) that contains the text of 50,000 movie reviews from the [Internet Movie Database](https://www.imdb.com/). These are split into 25,000 reviews for training and 25,000 reviews for testing. The training and testing sets are *balanced*, meaning they contain an equal number of positive and negative reviews. 

This notebook uses [`tf.keras`](https://www.tensorflow.org/guide/keras), a high-level API to build and train models in TensorFlow, and [`tensorflow_hub`](https://www.tensorflow.org/hub), a library for loading trained models from [TFHub](https://tfhub.dev) in a single line of code. For a more advanced text classification tutorial using `tf.keras`, see the [MLCC Text Classification Guide](https://developers.google.com/machine-learning/guides/text-classification/).

In [3]:
!pip install -q tensorflow-hub

In [4]:
import os
import numpy as np

import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.list_physical_devices("GPU") else "NOT AVAILABLE")

Version:  2.4.1
Eager mode:  True
Hub version:  0.11.0


GPU is available


## Download the IMDB dataset

The IMDB dataset is available on [imdb reviews](https://www.tensorflow.org/datasets/catalog/imdb_reviews) or on [TensorFlow datasets](https://www.tensorflow.org/datasets). The following code downloads the IMDB dataset to your machine (or the colab runtime):

In [5]:
# Split the training set into 60% and 40%, so we'll end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

## Explore the data 

Let's take a moment to understand the format of the data. Each example is a sentence representing the movie review and a corresponding label. The sentence is not preprocessed in any way. The label is an integer value of either 0 or 1, where 0 is a negative review, and 1 is a positive review.

Let's print first 10 examples.

In [6]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

Let's also print the first 10 labels.

In [7]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## Build the model

The neural network is created by stacking layers—this requires three main architectural decisions:

* How to represent the text?
* How many layers to use in the model?
* How many *hidden units* to use for each layer?

In this example, the input data consists of sentences. The labels to predict are either 0 or 1.

One way to represent the text is to convert sentences into embeddings vectors. We can use a pre-trained text embedding as the first layer, which will have three advantages:

*   we don't have to worry about text preprocessing,
*   we can benefit from transfer learning,
*   the embedding has a fixed size, so it's simpler to process.

For this example we will use a **pre-trained text embedding model** from [TensorFlow Hub](https://tfhub.dev) called [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2).

There are many other pre-trained text embeddings from TFHub that can be used in this tutorial:

* [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2) - trained with the same NNLM architecture on the same data as [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2), but with a larger embedding dimension. Larger dimensional embeddings can improve on your task but it may take longer to train your model.
* [google/nnlm-en-dim128-with-normalization/2](https://tfhub.dev/google/nnlm-en-dim128-with-normalization/2) - the same as [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2), but with additional text normalization such as removing punctuation. This can help if the text in your task contains additional characters or punctuation.
* [google/universal-sentence-encoder/4](https://tfhub.dev/google/universal-sentence-encoder/4) - a much larger model yielding 512 dimensional embeddings trained with a deep averaging network (DAN) encoder.

And many more! Find more [text embedding models](https://tfhub.dev/s?module-type=text-embedding) on TFHub.

Let's first create a Keras layer that uses a TensorFlow Hub model to embed the sentences, and try it out on a couple of input examples. Note that no matter the length of the input text, the output shape of the embeddings is: `(num_examples, embedding_dimension)`.

In [8]:
embedding = "https://tfhub.dev/google/nnlm-en-dim50/2"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 50), dtype=float32, numpy=
array([[ 0.5423195 , -0.0119017 ,  0.06337538,  0.06862972, -0.16776837,
        -0.10581174,  0.16865303, -0.04998824, -0.31148055,  0.07910346,
         0.15442263,  0.01488662,  0.03930153,  0.19772711, -0.12215476,
        -0.04120981, -0.2704109 , -0.21922152,  0.26517662, -0.80739075,
         0.25833532, -0.3100421 ,  0.28683215,  0.1943387 , -0.29036492,
         0.03862849, -0.7844411 , -0.0479324 ,  0.4110299 , -0.36388892,
        -0.58034706,  0.30269456,  0.3630897 , -0.15227164, -0.44391504,
         0.19462997,  0.19528408,  0.05666234,  0.2890704 , -0.28468323,
        -0.00531206,  0.0571938 , -0.3201318 , -0.04418665, -0.08550783,
        -0.55847436, -0.23336391, -0.20782952, -0.03543064, -0.17533456],
       [ 0.56338924, -0.12339553, -0.10862679,  0.7753425 , -0.07667089,
        -0.15752277,  0.01872335, -0.08169781, -0.3521876 ,  0.4637341 ,
        -0.08492756,  0.07166859, -0.00670817,  0.12686075, -0.19326553,
 

Let's now build the full model:

In [9]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 50)                48190600  
_________________________________________________________________
dense (Dense)                (None, 16)                816       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 48,191,433
Trainable params: 48,191,433
Non-trainable params: 0
_________________________________________________________________


The layers are stacked sequentially to build the classifier:

1. The first layer is a TensorFlow Hub layer. This layer uses a pre-trained Saved Model to map a sentence into its embedding vector. The pre-trained text embedding model that we are using ([google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2)) splits the sentence into tokens, embeds each token and then combines the embedding. The resulting dimensions are: `(num_examples, embedding_dimension)`. For this NNLM model, the `embedding_dimension` is 50.
2. This fixed-length output vector is piped through a fully-connected (`Dense`) layer with 16 hidden units.
3. The last layer is densely connected with a single output node.

Let's compile the model.

### Loss function and optimizer

A model needs a loss function and an optimizer for training. Since this is a binary classification problem and the model outputs logits (a single-unit layer with a linear activation), we'll use the `binary_crossentropy` loss function.

This isn't the only choice for a loss function, you could, for instance, choose `mean_squared_error`. But, generally, `binary_crossentropy` is better for dealing with probabilities—it measures the "distance" between probability distributions, or in our case, between the ground-truth distribution and the predictions.

Later, when we are exploring regression problems (say, to predict the price of a house), we will see how to use another loss function called mean squared error.

Now, configure the model to use an optimizer and a loss function:

In [10]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## Train the model

Train the model for 10 epochs in mini-batches of 512 samples. This is 10 iterations over all samples in the `x_train` and `y_train` tensors. While training, monitor the model's loss and accuracy on the 10,000 samples from the validation set:

In [11]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=10,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/10


 1/30 [>.............................] - ETA: 39s - loss: 0.6929 - accuracy: 0.5020

 2/30 [=>............................] - ETA: 1s - loss: 0.6911 - accuracy: 0.5098 

 4/30 [===>..........................] - ETA: 1s - loss: 0.6877 - accuracy: 0.5177

 6/30 [=====>........................] - ETA: 1s - loss: 0.6849 - accuracy: 0.5196

 8/30 [=======>......................] - ETA: 1s - loss: 0.6822 - accuracy: 0.5196

10/30 [=========>....................] - ETA: 0s - loss: 0.6796 - accuracy: 0.5202

12/30 [===========>..................] - ETA: 0s - loss: 0.6773 - accuracy: 0.5201

14/30 [=============>................] - ETA: 0s - loss: 0.6750 - accuracy: 0.5209

16/30 [===============>..............] - ETA: 0s - loss: 0.6727 - accuracy: 0.5223

18/30 [=================>............] - ETA: 0s - loss: 0.6706 - accuracy: 0.5245

20/30 [===================>..........] - ETA: 0s - loss: 0.6686 - accuracy: 0.5270

22/30 [=====================>........] - ETA: 0s - loss: 0.6665 - accuracy: 0.5298

24/30 [=======================>......] - ETA: 0s - loss: 0.6645 - accuracy: 0.5326

26/30 [=========================>....] - ETA: 0s - loss: 0.6624 - accuracy: 0.5352

28/30 [===========================>..] - ETA: 0s - loss: 0.6604 - accuracy: 0.5375

30/30 [==============================] - ETA: 0s - loss: 0.6584 - accuracy: 0.5398

30/30 [==============================] - 4s 92ms/step - loss: 0.6575 - accuracy: 0.5409 - val_loss: 0.5655 - val_accuracy: 0.6547


Epoch 2/10


 1/30 [>.............................] - ETA: 24s - loss: 0.5503 - accuracy: 0.6797

 3/30 [==>...........................] - ETA: 1s - loss: 0.5474 - accuracy: 0.6836 

 5/30 [====>.........................] - ETA: 1s - loss: 0.5448 - accuracy: 0.6901

 7/30 [======>.......................] - ETA: 0s - loss: 0.5427 - accuracy: 0.6952

 9/30 [========>.....................] - ETA: 0s - loss: 0.5402 - accuracy: 0.7008

11/30 [==========>...................] - ETA: 0s - loss: 0.5382 - accuracy: 0.7043

13/30 [============>.................] - ETA: 0s - loss: 0.5362 - accuracy: 0.7067

15/30 [==============>...............] - ETA: 0s - loss: 0.5342 - accuracy: 0.7090

17/30 [================>.............] - ETA: 0s - loss: 0.5321 - accuracy: 0.7110

19/30 [==================>...........] - ETA: 0s - loss: 0.5299 - accuracy: 0.7131

21/30 [====================>.........] - ETA: 0s - loss: 0.5277 - accuracy: 0.7152

23/30 [======================>.......] - ETA: 0s - loss: 0.5255 - accuracy: 0.7174

25/30 [========================>.....] - ETA: 0s - loss: 0.5234 - accuracy: 0.7195

27/30 [==========================>...] - ETA: 0s - loss: 0.5214 - accuracy: 0.7217

29/30 [============================>.] - ETA: 0s - loss: 0.5194 - accuracy: 0.7239

30/30 [==============================] - 3s 79ms/step - loss: 0.5175 - accuracy: 0.7258 - val_loss: 0.4537 - val_accuracy: 0.7627


Epoch 3/10


 1/30 [>.............................] - ETA: 24s - loss: 0.3993 - accuracy: 0.8008

 3/30 [==>...........................] - ETA: 1s - loss: 0.4032 - accuracy: 0.8070 

 5/30 [====>.........................] - ETA: 1s - loss: 0.4023 - accuracy: 0.8115

 7/30 [======>.......................] - ETA: 0s - loss: 0.3995 - accuracy: 0.8156

 9/30 [========>.....................] - ETA: 0s - loss: 0.3980 - accuracy: 0.8186

11/30 [==========>...................] - ETA: 0s - loss: 0.3962 - accuracy: 0.8220

13/30 [============>.................] - ETA: 0s - loss: 0.3945 - accuracy: 0.8249

15/30 [==============>...............] - ETA: 0s - loss: 0.3930 - accuracy: 0.8271

17/30 [================>.............] - ETA: 0s - loss: 0.3913 - accuracy: 0.8288

19/30 [==================>...........] - ETA: 0s - loss: 0.3898 - accuracy: 0.8301

21/30 [====================>.........] - ETA: 0s - loss: 0.3884 - accuracy: 0.8312

23/30 [======================>.......] - ETA: 0s - loss: 0.3869 - accuracy: 0.8324

25/30 [========================>.....] - ETA: 0s - loss: 0.3855 - accuracy: 0.8336

27/30 [==========================>...] - ETA: 0s - loss: 0.3839 - accuracy: 0.8348

29/30 [============================>.] - ETA: 0s - loss: 0.3825 - accuracy: 0.8359

30/30 [==============================] - 3s 78ms/step - loss: 0.3811 - accuracy: 0.8368 - val_loss: 0.3737 - val_accuracy: 0.8331


Epoch 4/10


 1/30 [>.............................] - ETA: 24s - loss: 0.2907 - accuracy: 0.8848

 3/30 [==>...........................] - ETA: 1s - loss: 0.2964 - accuracy: 0.8862 

 5/30 [====>.........................] - ETA: 1s - loss: 0.2951 - accuracy: 0.8861

 7/30 [======>.......................] - ETA: 1s - loss: 0.2930 - accuracy: 0.8868

 9/30 [========>.....................] - ETA: 0s - loss: 0.2917 - accuracy: 0.8872

11/30 [==========>...................] - ETA: 0s - loss: 0.2905 - accuracy: 0.8878

13/30 [============>.................] - ETA: 0s - loss: 0.2892 - accuracy: 0.8886

15/30 [==============>...............] - ETA: 0s - loss: 0.2877 - accuracy: 0.8894

17/30 [================>.............] - ETA: 0s - loss: 0.2863 - accuracy: 0.8902

19/30 [==================>...........] - ETA: 0s - loss: 0.2850 - accuracy: 0.8909

21/30 [====================>.........] - ETA: 0s - loss: 0.2838 - accuracy: 0.8916

23/30 [======================>.......] - ETA: 0s - loss: 0.2826 - accuracy: 0.8922

25/30 [========================>.....] - ETA: 0s - loss: 0.2815 - accuracy: 0.8928

27/30 [==========================>...] - ETA: 0s - loss: 0.2805 - accuracy: 0.8933

29/30 [============================>.] - ETA: 0s - loss: 0.2796 - accuracy: 0.8937

30/30 [==============================] - 3s 73ms/step - loss: 0.2788 - accuracy: 0.8941 - val_loss: 0.3330 - val_accuracy: 0.8527


Epoch 5/10


 1/30 [>.............................] - ETA: 19s - loss: 0.2034 - accuracy: 0.9238

 3/30 [==>...........................] - ETA: 1s - loss: 0.2037 - accuracy: 0.9248 

 5/30 [====>.........................] - ETA: 1s - loss: 0.2074 - accuracy: 0.9231

 7/30 [======>.......................] - ETA: 0s - loss: 0.2092 - accuracy: 0.9237

 9/30 [========>.....................] - ETA: 0s - loss: 0.2096 - accuracy: 0.9243

11/30 [==========>...................] - ETA: 0s - loss: 0.2098 - accuracy: 0.9247

13/30 [============>.................] - ETA: 0s - loss: 0.2096 - accuracy: 0.9253

15/30 [==============>...............] - ETA: 0s - loss: 0.2093 - accuracy: 0.9257

17/30 [================>.............] - ETA: 0s - loss: 0.2089 - accuracy: 0.9260

19/30 [==================>...........] - ETA: 0s - loss: 0.2085 - accuracy: 0.9263

21/30 [====================>.........] - ETA: 0s - loss: 0.2080 - accuracy: 0.9265

23/30 [======================>.......] - ETA: 0s - loss: 0.2076 - accuracy: 0.9267

25/30 [========================>.....] - ETA: 0s - loss: 0.2071 - accuracy: 0.9268

27/30 [==========================>...] - ETA: 0s - loss: 0.2066 - accuracy: 0.9270

29/30 [============================>.] - ETA: 0s - loss: 0.2062 - accuracy: 0.9273

30/30 [==============================] - 3s 77ms/step - loss: 0.2057 - accuracy: 0.9275 - val_loss: 0.3162 - val_accuracy: 0.8553


Epoch 6/10


 1/30 [>.............................] - ETA: 19s - loss: 0.1490 - accuracy: 0.9414

 3/30 [==>...........................] - ETA: 1s - loss: 0.1572 - accuracy: 0.9374 

 5/30 [====>.........................] - ETA: 1s - loss: 0.1583 - accuracy: 0.9382

 7/30 [======>.......................] - ETA: 0s - loss: 0.1596 - accuracy: 0.9391

 9/30 [========>.....................] - ETA: 0s - loss: 0.1595 - accuracy: 0.9408

11/30 [==========>...................] - ETA: 0s - loss: 0.1593 - accuracy: 0.9423

13/30 [============>.................] - ETA: 0s - loss: 0.1587 - accuracy: 0.9437

15/30 [==============>...............] - ETA: 0s - loss: 0.1583 - accuracy: 0.9446

17/30 [================>.............] - ETA: 0s - loss: 0.1578 - accuracy: 0.9454

19/30 [==================>...........] - ETA: 0s - loss: 0.1573 - accuracy: 0.9461

21/30 [====================>.........] - ETA: 0s - loss: 0.1569 - accuracy: 0.9467

23/30 [======================>.......] - ETA: 0s - loss: 0.1565 - accuracy: 0.9472

25/30 [========================>.....] - ETA: 0s - loss: 0.1560 - accuracy: 0.9476

27/30 [==========================>...] - ETA: 0s - loss: 0.1557 - accuracy: 0.9480

29/30 [============================>.] - ETA: 0s - loss: 0.1553 - accuracy: 0.9483

30/30 [==============================] - 3s 80ms/step - loss: 0.1549 - accuracy: 0.9486 - val_loss: 0.3063 - val_accuracy: 0.8661


Epoch 7/10


 1/30 [>.............................] - ETA: 20s - loss: 0.1165 - accuracy: 0.9727

 3/30 [==>...........................] - ETA: 1s - loss: 0.1178 - accuracy: 0.9712 

 5/30 [====>.........................] - ETA: 1s - loss: 0.1173 - accuracy: 0.9707

 7/30 [======>.......................] - ETA: 0s - loss: 0.1167 - accuracy: 0.9705

 9/30 [========>.....................] - ETA: 0s - loss: 0.1162 - accuracy: 0.9707

11/30 [==========>...................] - ETA: 0s - loss: 0.1158 - accuracy: 0.9707

13/30 [============>.................] - ETA: 0s - loss: 0.1154 - accuracy: 0.9708

15/30 [==============>...............] - ETA: 0s - loss: 0.1151 - accuracy: 0.9709

17/30 [================>.............] - ETA: 0s - loss: 0.1147 - accuracy: 0.9709

19/30 [==================>...........] - ETA: 0s - loss: 0.1143 - accuracy: 0.9710

21/30 [====================>.........] - ETA: 0s - loss: 0.1140 - accuracy: 0.9710

23/30 [======================>.......] - ETA: 0s - loss: 0.1137 - accuracy: 0.9710

25/30 [========================>.....] - ETA: 0s - loss: 0.1135 - accuracy: 0.9710

27/30 [==========================>...] - ETA: 0s - loss: 0.1133 - accuracy: 0.9710

29/30 [============================>.] - ETA: 0s - loss: 0.1130 - accuracy: 0.9710

30/30 [==============================] - 3s 72ms/step - loss: 0.1128 - accuracy: 0.9709 - val_loss: 0.3075 - val_accuracy: 0.8710


Epoch 8/10


 1/30 [>.............................] - ETA: 19s - loss: 0.0784 - accuracy: 0.9863

 3/30 [==>...........................] - ETA: 1s - loss: 0.0821 - accuracy: 0.9834 

 5/30 [====>.........................] - ETA: 1s - loss: 0.0836 - accuracy: 0.9826

 7/30 [======>.......................] - ETA: 0s - loss: 0.0843 - accuracy: 0.9821

 9/30 [========>.....................] - ETA: 0s - loss: 0.0844 - accuracy: 0.9821

11/30 [==========>...................] - ETA: 0s - loss: 0.0845 - accuracy: 0.9819

13/30 [============>.................] - ETA: 0s - loss: 0.0845 - accuracy: 0.9817

15/30 [==============>...............] - ETA: 0s - loss: 0.0844 - accuracy: 0.9817

17/30 [================>.............] - ETA: 0s - loss: 0.0841 - accuracy: 0.9816

19/30 [==================>...........] - ETA: 0s - loss: 0.0839 - accuracy: 0.9816

21/30 [====================>.........] - ETA: 0s - loss: 0.0836 - accuracy: 0.9817

23/30 [======================>.......] - ETA: 0s - loss: 0.0833 - accuracy: 0.9817

25/30 [========================>.....] - ETA: 0s - loss: 0.0831 - accuracy: 0.9817

27/30 [==========================>...] - ETA: 0s - loss: 0.0828 - accuracy: 0.9818

29/30 [============================>.] - ETA: 0s - loss: 0.0826 - accuracy: 0.9818

30/30 [==============================] - 3s 78ms/step - loss: 0.0824 - accuracy: 0.9818 - val_loss: 0.3171 - val_accuracy: 0.8723


Epoch 9/10


 1/30 [>.............................] - ETA: 19s - loss: 0.0550 - accuracy: 0.9902

 3/30 [==>...........................] - ETA: 1s - loss: 0.0568 - accuracy: 0.9893 

 5/30 [====>.........................] - ETA: 1s - loss: 0.0570 - accuracy: 0.9895

 7/30 [======>.......................] - ETA: 0s - loss: 0.0571 - accuracy: 0.9896

 9/30 [========>.....................] - ETA: 0s - loss: 0.0569 - accuracy: 0.9898

11/30 [==========>...................] - ETA: 0s - loss: 0.0570 - accuracy: 0.9897

13/30 [============>.................] - ETA: 0s - loss: 0.0572 - accuracy: 0.9896

15/30 [==============>...............] - ETA: 0s - loss: 0.0573 - accuracy: 0.9896

17/30 [================>.............] - ETA: 0s - loss: 0.0574 - accuracy: 0.9896

19/30 [==================>...........] - ETA: 0s - loss: 0.0574 - accuracy: 0.9897

21/30 [====================>.........] - ETA: 0s - loss: 0.0574 - accuracy: 0.9896

23/30 [======================>.......] - ETA: 0s - loss: 0.0573 - accuracy: 0.9896

25/30 [========================>.....] - ETA: 0s - loss: 0.0573 - accuracy: 0.9896

27/30 [==========================>...] - ETA: 0s - loss: 0.0572 - accuracy: 0.9896

29/30 [============================>.] - ETA: 0s - loss: 0.0572 - accuracy: 0.9896

30/30 [==============================] - 3s 73ms/step - loss: 0.0571 - accuracy: 0.9896 - val_loss: 0.3260 - val_accuracy: 0.8676


Epoch 10/10


 1/30 [>.............................] - ETA: 53s - loss: 0.0436 - accuracy: 0.9941

 3/30 [==>...........................] - ETA: 1s - loss: 0.0423 - accuracy: 0.9954 

 5/30 [====>.........................] - ETA: 1s - loss: 0.0418 - accuracy: 0.9954

 7/30 [======>.......................] - ETA: 0s - loss: 0.0417 - accuracy: 0.9954

 9/30 [========>.....................] - ETA: 0s - loss: 0.0418 - accuracy: 0.9953

11/30 [==========>...................] - ETA: 0s - loss: 0.0418 - accuracy: 0.9954

13/30 [============>.................] - ETA: 0s - loss: 0.0417 - accuracy: 0.9954

15/30 [==============>...............] - ETA: 0s - loss: 0.0416 - accuracy: 0.9954

17/30 [================>.............] - ETA: 0s - loss: 0.0416 - accuracy: 0.9954

19/30 [==================>...........] - ETA: 0s - loss: 0.0415 - accuracy: 0.9954

21/30 [====================>.........] - ETA: 0s - loss: 0.0415 - accuracy: 0.9954

23/30 [======================>.......] - ETA: 0s - loss: 0.0414 - accuracy: 0.9953

25/30 [========================>.....] - ETA: 0s - loss: 0.0414 - accuracy: 0.9953

27/30 [==========================>...] - ETA: 0s - loss: 0.0413 - accuracy: 0.9953

29/30 [============================>.] - ETA: 0s - loss: 0.0412 - accuracy: 0.9953

30/30 [==============================] - 4s 75ms/step - loss: 0.0412 - accuracy: 0.9952 - val_loss: 0.3405 - val_accuracy: 0.8711


## Evaluate the model

And let's see how the model performs. Two values will be returned. Loss (a number which represents our error, lower values are better), and accuracy.

In [12]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 2s - loss: 0.3628 - accuracy: 0.8556


loss: 0.363
accuracy: 0.856


This fairly naive approach achieves an accuracy of about 87%. With more advanced approaches, the model should get closer to 95%.

## Further reading

* For a more general way to work with string inputs and for a more detailed analysis of the progress of accuracy and loss during training, see the [Text classification with preprocessed text](./text_classification.ipynb) tutorial.
* Try out more [text-related tutorials](https://www.tensorflow.org/hub/tutorials#text-related-tutorials) using trained models from TFHub.